service/raft/raft_group0.cc

```diff
@@ -949,7 +949,11 @@ with_timeout(abort_source& as, db::timeout_clock::duration d, F&& fun) {
     // FIXME: using lambda as workaround for clang bug #50345 (miscompiling coroutine templates).
     auto impl = [] (abort_source& as, db::timeout_clock::duration d, F&& fun) -> future_t {
         abort_source timeout_src;
-        auto sub = as.subscribe([&timeout_src] () noexcept { timeout_src.request_abort(); });
+        auto sub = as.subscribe([&timeout_src] () noexcept {
+            if (!timeout_src.abort_requested()) {
+                timeout_src.request_abort();
+            }
+        });
```

Comment:
I think I've seen this pattern before. Maybe we should relax abort_source checks here. Not suggesting to block this patch.

Reply:
I've also seen the pattern of linking abort_sources together to form scopes. Perhaps abort_source should support it directly (by capturing a wider-scope abort_source in its constructor).

Reply:
> I think I've seen this pattern before. Maybe we should relax abort_source checks here. Not suggesting to block this patch.

It think that the point of double-abort protection came from a concern that the abort_source itself may be destoryed right after abort is requested, although that's rarely the case as the abort_source lives longer than the resource it's controlling.

It might be safer to define a shared abort source (that could carry an `abort_source` member) for these use cases where it may be aborted from multiple sources and tolerate multiple closures in its `request_abort` method by calling the member's request_abort only once.

Reply:
> I've also seen the pattern of linking abort_sources together to form scopes. Perhaps abort_source should support it directly (by capturing a wider-scope abort_source in its constructor).

Yes, we have that in the task_manager where the service subscribes to the global abort source, then it has modules, each with an abort_source, and tasks in each module, all subscribing to the module's abort source, while each task has an abort source so it could be aborted independently.

Aborting a task automatically also aborts all its descendants recursively.

So abstracring this abort tree in a generic seems very appealing.

Cc @Deexie

Reply:
I opened a draft https://github.com/scylladb/seastar/pull/1425 with that. It's still work in progress and I haven't tested it yet.

It'd be nice if you took a look whether I got your idea properly.